client/src/components/TimeSetter.tsx

```diff
@@ -1,22 +1,32 @@
-import { useState, FC } from "react";
+import { FC, useContext } from "react";
 import { Typography, Card, CardContent, TextField } from "@mui/material";
 import AdapterDateFns from "@mui/lab/AdapterDateFns";
 import { LocalizationProvider, TimePicker } from "@mui/lab";
+import { FormContext } from "../pages/bookingPage";
 
 type Props = {
   title: string;
 };
+
 export const TimeSetter: FC<Props> = ({ title }) => {
-  const [timeValue, setTimeValue] = useState<Date | null>(null);
+  const {
+    startTime: [startTimeValue, setStartTimeValue],
+  } = useContext(FormContext);
+  const {
+    endTime: [endTimeValue, setEndTimeValue],
+  } = useContext(FormContext);
   return (
     <Card>
       <CardContent>
         <Typography variant="body2">{title}</Typography>
         <LocalizationProvider dateAdapter={AdapterDateFns}>
           <TimePicker
-            value={timeValue}
+            value={title.includes("Start") ? startTimeValue : endTimeValue}
             onChange={(newValue) => {
-              setTimeValue(newValue);
+              if (newValue) {
+                if (title.includes("Start")) setStartTimeValue(newValue);
```

Comment:
Ah yes, this cunning trick.

It's not wrong, but it can be more explicit and slightly neater. I would advice adding a prop to switch specifically between these 2 options, because you can't have any other values here and title can be any string!

This is still not the best way to do this, maybe I will think about that later :)

So you could do this:

```
type Props = {
  title: string;
  type: "start" | "end";
}

...
// do not nest too many if statements.
if(!newValue) {
  return;
}
// avoid the if syntax without {} for readability
if(type === "start") {
   setStartTimeValue(...
}
```

Reply:
OK, this is still a bit better way. Can you identify why?

```
import { FC, useContext } from "react";
import { Typography, Card, CardContent, TextField } from "@mui/material";
import AdapterDateFns from "@mui/lab/AdapterDateFns";
import { LocalizationProvider, TimePicker } from "@mui/lab";
import { FormContext } from "../pages/bookingPage";

type Props = {
  title: string;
  type: "startTime" | "endTime";
};

export const TimeSetter: FC<Props> = ({ title, type }) => {
  // const {
  //   startTime: [startTimeValue, setStartTimeValue],
  // } = useContext(FormContext);
  // const {
  //   endTime: [endTimeValue, setEndTimeValue],
  // } = useContext(FormContext);
  const {
    [type]: [timeValue, setTimeValue],
  } = useContext(FormContext);
  return (
    <Card>
      <CardContent>
        <Typography variant="body2">{title}</Typography>
        <LocalizationProvider dateAdapter={AdapterDateFns}>
          <TimePicker
            value="TODO"
            // value={type.includes("Start") ? startTimeValue : endTimeValue}
            onChange={(newValue) => {
              if (newValue) {
                setTimeValue(newValue);
                // if (title.includes("Start")) setStartTimeValue(newValue);
                // if (title.includes("End")) setEndTimeValue(newValue);
              }
            }}
            renderInput={(params) => <TextField {...params} />}
          />
        </LocalizationProvider>
      </CardContent>
    </Card>
  );
};

```